lxd/instance/instance_utils.go

```diff
@@ -128,6 +128,45 @@ func ValidConfig(sysOS *sys.OS, config map[string]string, expanded bool, instanc
 }
 
 func validConfigKey(os *sys.OS, key string, value string, instanceType instancetype.Type) error {
+	if instanceType != instancetype.Any &&
+		!shared.StringHasPrefix(key, instancetype.ConfigKeyPrefixesAny...) &&
+		!strings.HasPrefix(key, instancetype.ConfigVolatilePrefix) {
+		// Determine valid keys for instance type.
+		var validKeys map[string]func(string) error
+		switch instanceType {
+		case instancetype.VM:
+			if shared.StringHasPrefix(key, instancetype.ConfigKeyPrefixesContainer...) {
+				return fmt.Errorf("%s isn't supported for VMs", key)
+			}
+
+			validKeys = instancetype.InstanceConfigKeysVM
+		case instancetype.Container:
+			validKeys = instancetype.InstanceConfigKeysContainer
+		}
+
+		// Check if the key is valid.
+		if validKeys != nil {
+			_, exists := validKeys[key]
+			_, existsAny := instancetype.InstanceConfigKeysAny[key]
+			if !(exists || existsAny) {
+				return fmt.Errorf("%s isn't supported for %ss", key, instanceType)
+			}
+		}
+	}
+
+	// Check if the key requires a subkey.
+	var matchedPrefix string
+	for _, prefix := range append(instancetype.ConfigKeyPrefixesAny, instancetype.ConfigKeyPrefixesContainer...) {
+		if strings.HasPrefix(key, prefix) {
+			matchedPrefix = prefix
+			break
+		}
+	}
+
+	if matchedPrefix != "" && len(key) <= len(matchedPrefix) {
+		return fmt.Errorf("%s requires a subkey", key)
+	}
```

Comment:
I think this simplification would work
```suggestion
	for _, prefix := range append(instancetype.ConfigKeyPrefixesAny, instancetype.ConfigKeyPrefixesContainer...) {
                // Reject the key if it matches a known prefix as it indicates the subkey was omitted.
		if key == prefix {
			return fmt.Errorf("%s requires a subkey", key)
		}
	}
```